File: homework1/hw1.m
load hw1_15_train.dat
load hw1_18_train.dat
load hw1_18_test.dat
pkg load statistics

printf('Run pla with order\n');
[w times] = pla(hw1_15_train(:, 1:4), hw1_15_train(:, 5), 1, 'order')
pause

printf('Run pla with random for 2000 times\n');
times_array = zeros(1, 2000);
times_sum = 0;
for n = 1:2000
    printf('.');
    rand('seed', n);
    [w times] = pla(hw1_15_train(:, 1:4), hw1_15_train(:, 5), 1, 'random');
    times_sum += times;
    times_array(n) = times;
end
printf('\n');
printf('Average times of pla is %d\n', times_sum / 2000);
pause
hist(times_array);

printf('Run pocket with order\n');
[w err] = pocket(hw1_18_train(:, 1:4), hw1_18_train(:, 5), 1, 'order')
pause

printf('Run pocket with random for 2000 times\n');
err_array = zeros(1, 2000);
err_sum = 0;
for n = 1:2000
    printf('.');
    rand('seed', n);
    [w err] = pocket(hw1_18_train(:, 1:4), hw1_18_train(:, 5), 1, 'random');
    x = hw1_18_test(:, 1:4);
    X = [ones(size(x, 1), 1) x];
    y = hw1_18_test(:, 5);
    err = sum(sign(X * w') != y) / size(X, 1);
    err_sum += err;
    err_array(n) = err;
end
printf('\n');
printf('Average error of pocket is %d\n', err_sum / 2000);
pause
hist(err_array);

File: homework1/hw1.R
source("common.R", echo = F)
source("homework1/pla.R", echo = F)
source("homework1/pocket.R", echo = F)
hw1_15_train <- as.matrix(read.table("homework1/hw1_15_train.dat", stringsAsFactors = T))
hw1_18_train <- as.matrix(read.table("homework1/hw1_18_train.dat", stringsAsFactors = T))
hw1_18_test <- as.matrix(read.table("homework1/hw1_18_test.dat", stringsAsFactors = T))


result <- pla(hw1_15_train[, 1:4], v(hw1_15_train, 5), 1, "order")
message(paste("Run pla with order w is", result[1], "times is", result[[2]]))
pause()

message('Run pla with random for 2000 times')
times_array <- rep(0, 2000)
times_sum <- 0
for (i in 1:2000) {
  cat("=")
  set.seed(i)
  result <- pla(hw1_15_train[, 1:4], v(hw1_15_train, 5), 1, "random")
  times_sum <- times_sum + result[[2]]
  times_array[i] <- result[[2]]
}
message("Average times of pla is ", times_sum / 2000)
hist(times_array)
pause()

result <- pocket(hw1_15_train[, 1:4], v(hw1_15_train, 5), 1, "order")
message(paste("Run pocket with order, w is", result[1], "times is", result[[2]]))
pause()

message("Run pocket with random for 2000 times")
err_array <- rep(0, 2000)
err_sum <- 0
for (i in 1:2000) {
  cat("=")
  set.seed(i)
  result <- pocket(hw1_18_train[, 1:4], v(hw1_18_train, 5), 1, "random")
  x <- hw1_18_test[, 1:4]
  X <- cbind(as.matrix(rep(1, nrow(x))), x)
  y <- v(hw1_18_test, 5)
  w <- result[[1]]
  err <- sum(sign(X %*% w) != y) / nrow(x)
  err_sum <- err_sum + err
  err_array[i] <- err
}
message("Average error of pocket is ", err_sum / 2000)
pause()

hist(err_array)
